runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/ChannelCache.java

```diff
@@ -0,0 +1,114 @@
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.beam.runners.dataflow.worker.status.StatusDataProvider;
+import org.apache.beam.runners.dataflow.worker.windmill.WindmillServiceAddress;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ManagedChannel;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Cache of gRPC channels for {@link WindmillServiceAddress}. Follows <a
+ * href=https://grpc.io/docs/guides/performance/#java>gRPC recommendations</a> for re-using channels
+ * when possible.
+ *
+ * @implNote Backed by {@link LoadingCache} which is thread-safe.
+ */
+@ThreadSafe
+public final class ChannelCache implements StatusDataProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(ChannelCache.class);
+  private final LoadingCache<WindmillServiceAddress, ManagedChannel> channelCache;
+
+  private ChannelCache(
+      Function<WindmillServiceAddress, ManagedChannel> channelFactory,
+      RemovalListener<WindmillServiceAddress, ManagedChannel> onChannelRemoved) {
+    this.channelCache =
+        Caffeine.newBuilder().removalListener(onChannelRemoved).build(channelFactory::apply);
+  }
+
+  public static ChannelCache create(
+      Function<WindmillServiceAddress, ManagedChannel> channelFactory) {
+    return new ChannelCache(
+        channelFactory,
+        // Shutdown the channels as they get removed from the cache, so they do not leak.
+        (address, channel, cause) -> shutdownChannel(channel));
+  }
+
+  @VisibleForTesting
+  static ChannelCache forTesting(
+      Function<WindmillServiceAddress, ManagedChannel> channelFactory, Runnable onChannelShutdown) {
+    return new ChannelCache(
+        channelFactory,
+        // Shutdown the channels as they get removed from the cache, so they do not leak.
+        // Add hook for testing so that we don't have to sleep/wait for arbitrary time in test.
+        (address, channel, cause) -> {
+          shutdownChannel(channel);
+          onChannelShutdown.run();
+        });
+  }
+
+  private static void shutdownChannel(ManagedChannel channel) {
+    channel.shutdown();
+    try {
+      channel.awaitTermination(10, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      LOG.error("Couldn't close gRPC channel={}", channel, e);
+    }
+    channel.shutdownNow();
+  }
+
+  public ManagedChannel get(WindmillServiceAddress windmillServiceAddress) {
+    return channelCache.get(windmillServiceAddress);
+  }
+
+  public void remove(WindmillServiceAddress windmillServiceAddress) {
+    channelCache.invalidate(windmillServiceAddress);
+  }
+
+  public void clear() {
+    channelCache.invalidateAll();
+  }
+
+  @VisibleForTesting
+  boolean isEmpty() {
+    // Perform any pending removal/insert operations first.
+    channelCache.cleanUp();
```

Comment:
does this block for removal listeners if they are on executor? add comment

Reply:
yes this does block added comment